cli: be more diligent about reporting progress during decommissioning

knz opened the pull request:
Fixes #75454.

Release note (cli change): `cockroach node decommission` now reports
progress more frequently.

Release note (cli change): The `cockroach node decommission` CLI
utility is now able to report progress messages even if the table
`system.eventlog` is unavailable server-side.  (Note that the
decommission process is able to make progress even when
system.eventlog is unavailable regardless of this change; this note is
focusing specifically on the reporting of progress messages by the CLI
utility.)

knz committed:
cli: be more diligent about reporting progress during decommissioning

Prior to this change, we had two situations where the CLI tool
`cockroach node decommission` was causing surprising behavior:

- when the `Decommission` RPC was hanging, the CLI was also hanging
  and there was no indication of this.

  This commit fixes this by introducing a timeout, and retrying
  when the timeout is reached.

- when the `Decommission` RPC was failing to make progress (prior and
  new counts were equal) forever, the output was simply stopping and
  there was no way for the observer to get a periodic reminder of the
  count.

  This commit ensures that the deco progress table is printed at least
  once every 5 seconds, even when no progress is made.

Release note (cli change): `cockroach node decommission` now reports
progress more frequently.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75467)
<!-- Reviewable:end -->

knz commented:
@AlexTalks do you have an issue number I can link to this PR?

@tbg how do you feel about adding a test for this? I think the quality-of-life improvement makes this PR important, but the test infra needed to ascertain the behavior is large. Is it worth the time investment?

knz committed:
server: avoid holding up the decommission CLI if system.eventlog is unavailable

Release note (cli change): The `cockroach node decommission` CLI
utility is now able to report progress messages even if the table
`system.eventlog` is unavailable server-side.  (Note that the
decommission process is able to make progress even when
system.eventlog is unavailable regardless of this change; this note is
focusing specifically on the reporting of progress messages by the CLI
utility.)

tbg commented:
> how do you feel about adding a test for this? I think the quality-of-life improvement makes this PR important, but the test infra needed to ascertain the behavior is large. Is it worth the time investment?

I'm okay with just manual testing of this, but I also think that making this code testable would be useful. I think this could be achieved if we pulled out the code inside of the retry loop into a unit-testable function (that in particular receives the time since last response and the last response, and takes an interface for the decommission RPC. It is going to be a lot of work and whether it is justified depends on what the opportunity cost is.

cameronnunez reviewed (commented):
LGTM! Needs that conflict resolution though because of the recent patch to decommission.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75467#-:-MvLG66SAeYYqmwBcxgz:b-kex2wz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @knz)

___
*[pkg/cli/node.go, line 458 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75467#-MvLGW6-689GHTk8KTPs:-MvLGW6-689GHTk8KTPt:b-45g766) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c5a4827166e60fd79e68d4a5842ecc81a3e622e/pkg/cli/node.go#L458)):*
> ```Go
> 				fmt.Fprintln(stderr)
> 
> 				// If the deco RPC times out, onsider this to be a transient
> ```

nit: "consider this to be a transient"
___
*[pkg/server/server.go, line 2667 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75467#-MvLH66s49VlJvlXawbC:-MvLH66s49VlJvlXawbD:b-9a2643) ([raw file](https://github.com/cockroachdb/cockroach/blob/68203eaa3b489c92f39f38a0a375b1b55ddb182b/pkg/server/server.go#L2667)):*
> ```Go
> 						int32(s.NodeID()), /* reporting ID: the node where the event is logged */
> 						sql.LogToSystemTable|sql.LogToDevChannelIfVerbose, /* we already call log.StructuredEvent above */
> 						int32(nodeID), /* target ID: the node that we wee a membership change for */
> ```

nit: pre-existing comment misprint "target ID: the node that we see a membership change for"


<!-- Sent from Reviewable.io -->

knz commented:
I'm not able to bring this PR to completion. @AlexTalks do you want to pick it up, or maybe take a different/new approach?

AlexTalks reviewed (commented):
Commenting in order to remove this from my pending reviews 😆
